zap/src/main/java/org/zaproxy/zap/extension/ascan/filters/GenericFilterUtility.java

```diff
@@ -0,0 +1,272 @@
+                        return FilterResult.NOT_FILTERED;
+                    }
+                    for (T value : nodeValues) {
+                        if (genericFilterData.getValues().contains(value)) {
+                            return FilterResult.NOT_FILTERED;
+                        }
+                    }
+
+                    return new FilterResult(
+                            Constant.messages.getString(
+                                    INCLUDE_FILTER_CRITERIA_MESSAGE_KEY,
+                                    new Object[] {filterType, genericFilterData.getValues()}));
+                case EXCLUDE:
+                    for (T value : nodeValues) {
+                        if (genericFilterData.getValues().contains(value)) {
+                            return new FilterResult(
+                                    Constant.messages.getString(
+                                            EXCLUDE_FILTER_CRITERIA_MESSAGE_KEY,
+                                            new Object[] {filterType, "[" + value + "]"}));
+                        }
+                    }
+                    return FilterResult.NOT_FILTERED;
+                default:
+                    return FilterResult.NOT_FILTERED;
+            }
+        }
+        return FilterResult.NOT_FILTERED;
+    }
+
+    /**
+     * this method will call {@code GenericFilterUtility#isFiltered(Collection, Collection, String)}
+     * method internally so basic requirement of equals and hasCode for the Classes holds same.
+     *
+     * @param <T> any object implementing hashCode and equals method
+     * @param genericFilterDataCollection Collection of data provided by setting filter criteria
+     * @param nodeValue data associated with the scanned node.
+     * @param filterType type of the filter ie Tag/Http Status code/Method etc
+     * @return FilterResult contains result after applying filter.
+     */
+    public static <T> FilterResult isFiltered(
+            Collection<GenericFilterData<T>> genericFilterDataCollection,
+            T nodeValue,
+            String filterType) {
+        Set<T> nodeValues = new LinkedHashSet<>();
+        nodeValues.add(nodeValue);
+        return isFiltered(genericFilterDataCollection, nodeValues, filterType);
+    }
+
+    /**
+     * This method is used to create {@link TagScanFilter} instance using Tag List for Include and
+     * Exclude {@code FilterCriteria}
+     *
+     * @param incTagList
+     * @param excTagList
+     * @return List of ScanFilter
+     */
+    public static List<ScanFilter> getTagScanFilters(
+            List<String> incTagList, List<String> excTagList) {
+        List<ScanFilter> scanFilterList = new ArrayList<>();
+        if (!CollectionUtils.isEmpty(incTagList)) {
+            GenericFilterData<String> genericFilterData =
+                    new GenericFilterData<String>(FilterCriteria.INCLUDE, incTagList);
+            TagScanFilter tagScanFilter = new TagScanFilter();
```

Comment:
because there can be include as well exclude tag filter, eg user can choose to include {hidden tag} but want to exclude {json tag}

Reply:
I thought that `AbstractScanFilter` supported that... but yeah, reading the code it just handles one `FilterCriteria` at the same time (which means that the `FilterCriteria` should be in the filter not the data). IMO we should change `AbstractScanFilter` to properly handle the `FilterCriteria`.

Reply:
hmm ... oh you are saying to add filtercriteria to filter as it is responsibility of filter not the data. will do that change. but what ever the case is, we need to return multiple filters.

Reply:
If `AbstractScanFilter` takes into account the filter criteria we would just need to use one filter, since we could add both the include/exclude data. If we are going to move the filter criteria to the filter then the data could be moved as well.

Reply:
didn't got you fully, but in that case filter need to have both include data and exclude data which i think is not right as in case a new criteria comes then filter has to add a new List or collection to hold that data.

Reply:
It wouldn't need more lists/collections, the filter criteria already indicates what the list/data applies to.

Lets step back and talk about the current state and what it should look like.

Reply:
didn't got that point. can you please add the snippet for better understanding.

Reply:
My initial suggestion was to change the exiting logic to check all `GenericFilterData` instead of returning immediately.